File: Leetcode/cpp/Flatten Binary Tree to Linked List.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode *root) {
        if (!root) return;
        if (root -> left) {
            TreeNode* tn = rightmost(root -> left);
            tn -> right = root -> right;
            root -> right = root -> left;
            root -> left = NULL;
        }
        flatten(root -> right);
    }
    TreeNode* rightmost(TreeNode* root) {
        while (root && root -> right) {
            root = root -> right;
        }
        return root;
    }
};

File: Leetcode/python/Flatten Binary Tree to Linked List.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        if not root:
            return None
        rightmost_of_left = self.flatten(root.left)
        if rightmost_of_left:
           rightmost_of_left.right, root.left, root.right = root.right, None, root.left
        if root.right:
            return self.flatten(root.right)
        else:
            return root
